neuro_extras/main.py

```diff
@@ -201,6 +231,155 @@ def image() -> None:
+                logger.error(f"  Reason: {job.history.reason}")
+                logger.error(f"  Description: {job.history.description}")
+                exit_code = job.history.exit_code
+                if exit_code is None:
+                    exit_code = EX_PLATFORMERROR
+                sys.exit(exit_code)
+            else:
+                logger.info("Successfully copied data")
+    else:
+        TEMP_UNPACK_DIR.mkdir(parents=True, exist_ok=True)
+        tmp_dir_name = tempfile.mkdtemp(dir=str(TEMP_UNPACK_DIR))
+        tmp_dst_url = URL.build(path=(tmp_dir_name + "/"))
+
+        await _nonstorage_cp(source_url, tmp_dst_url)
+        source_url = tmp_dst_url
+        # for clarity
+        source_url_type = UrlType.LOCAL
+
+        # at this point source is always local
+        if unpack:
+            file = Path(source_url.path)
+            if file.is_dir():
+                file = list(file.glob("*"))[0]
+            suffixes = file.suffixes
+            if suffixes[-2:] == [".tar", ".gz"] or suffixes[-1] == ".tgz":
+                command = "tar"
+                args = ["zxvf", str(file), "-C", str(destination_url)]
+                click.echo(f"Running {command} {' '.join(args)}")
+                subprocess = await asyncio.create_subprocess_exec(command, *args)
+                returncode = await subprocess.wait()
+                if returncode != 0:
+                    raise click.ClickException("Cloud copy failed")
+            elif suffixes[-1] in (".zip", ".gz"):
+                # TODO: Implement gunzip
+                # gunzip -c file.gz > /THERE/file
+                raise NotImplementedError("Pure gzip is not supported yet")
+            else:
+                raise ValueError(f"Don't know how to unpack file with {file.name}")
+
+        # handle upload/rsync
+        await _nonstorage_cp(source_url, destination_url)
+
+
+async def _nonstorage_cp(source_url: URL, destination_url: URL) -> None:
+    if "s3" in (source_url.scheme, destination_url.scheme):
+        command = "aws"
+        args = ["s3", "cp", str(source_url), str(destination_url)]
+        if source_url.path.endswith("/"):
+            args.insert(2, "--recursive")
+    elif "gs" in (source_url.scheme, destination_url.scheme):
+        command = "gsutil"
+        args = ["-m", "cp", "-r", str(source_url), str(destination_url)]
+    elif source_url.scheme == "" and destination_url.scheme == "":
+        command = "rsync"
+        args = ["-avzh", "--progress", str(source_url), str(destination_url)]
+    else:
+        raise ValueError("Unknown cloud provider")
+    click.echo(f"Running {command} {' '.join(args)}")
+    subprocess = await asyncio.create_subprocess_exec(command, *args)
+    returncode = await subprocess.wait()
+    if returncode != 0:
+        raise click.ClickException("Cloud copy failed")
+
+
+@data.command("cp")
+@click.argument("source")
+@click.argument("destination")
```

Comment:
Important: `destination` is always considered to be a directory, so that:
```
$ neuro-extras data cp s3://data.neu.ro/bone-age-tiny.zip /tmp/data.zip
```
will create directory `/tmp/data.zip` and put there file `bone-age-tiny.zip`. 

Good solution: support POSIX-like `-T` operator (see `neuro cp --help` -- neuro supports this). 
Quick solution: rename parameter `destination` -> `destination_dir` and add help message explicitly stating this fact.